botocore/regions.py

```diff
@@ -63,8 +64,13 @@ def construct_endpoint(self, service_name, region_name, **kwargs):
                                          region_name, **kwargs)
 
         if endpoint is None:
-            raise UnknownEndpointError(service_name=service_name,
-                                       region_name=region_name)
+            if region_name is None:
+                # Raise a more specific error message that will give
+                # better guidance to the user what needs to happen.
+                raise NoRegionError()
+            else:
+                raise UnknownEndpointError(service_name=service_name,
```

Comment:
This is tested here: https://github.com/boto/botocore/pull/493/files#diff-e323551d51dbf48e6eb73d90a9d46809R338

You'll get this thrown if we don't resolve an endpoint, but we specified a region.  In practice this will happen if the heuristic does not have a catch all rule.